kmcd-prelude.md

```diff
@@ -144,6 +143,134 @@ module KMCD-PRELUDE
+    rule chooseString (I, ITEMS) => { ITEMS [ I modInt size(ITEMS) ] }:>String
+    rule chooseAddress(I, ITEMS) => { ITEMS [ I modInt size(ITEMS) ] }:>Address
+    rule chooseCDPID  (I, ITEMS) => { ITEMS [ I modInt size(ITEMS) ] }:>CDPID
+
+    syntax AdminStep ::= AddGenerator ( GenStep )
+ // ---------------------------------------------
+    rule <k> AddGenerator ( GSS ) => . ... </k>
+         <generator-next> I => I +Int 1 </generator-next>
+         <generators>
+           ...
+           ( .Bag
+          => <generator>
+               <generator-id> I </generator-id>
+               <generator-steps> GSS </generator-steps>
+             </generator>
+           )
+           ...
+         </generators>
+
+    syntax AdminStep ::= GenStep
+    syntax GenStep ::= "GenStep"
+                     | "GenStepLoad"
+                     | "GenStepReplace"
+ // -----------------------------------
+    rule <k> GenStep => GenStepLoad ~> GenStepReplace ... </k>
+         <random> I => randInt(I) </random>
+         <generator-next> N </generator-next>
+         <generator-current> _ => I modInt N </generator-current>
+         <generator-remainder> GSS => .GenStep </generator-remainder>
+         <generator-depth-bound> DB => decrement(DB) </generator-depth-bound>
+         <violation> false </violation>
+      requires DB =/=K 0
+
+    rule <k> GenStepLoad => GSS ... </k>
+         <generator-current> I </generator-current>
+         <generator>
+           <generator-id> I </generator-id>
+           <generator-steps> GSS => .GenStep </generator-steps>
+         </generator>
+
+    rule <k> GenStepReplace => . ... </k>
+         <generator-remainder> GSS => .GenStep </generator-remainder>
+         <generator-current> I </generator-current>
+         <generator>
+           <generator-id> I </generator-id>
+           <generator-steps> _ => GSS </generator-steps>
+         </generator>
+
+    syntax AdminStep ::= LogGen ( MCDStep )
+    syntax Event ::= GenStep ( MCDStep )
+                   | GenStepFailed ( GenStep )
+ // ------------------------------------------
+    rule <k> LogGen(MCDSTEP) => MCDSTEP ... </k>
+         <events> ... (.List => ListItem(GenStep(MCDSTEP))) </events>
+
+    rule <k> GS => . ... </k>
+         <events> ... (.List => ListItem(GenStepFailed(GS))) </events>
+      [owise]
+
+    syntax GenStep ::= ".GenStep"
+                     | GenStep DepthBound
+                     > GenStep "|" GenStep [left]
+                     > GenStep ";" GenStep [left]
+ // ---------------------------------------------
+    rule <k> .GenStep => . ... </k>
+
+    rule <k> .GenStep ; GSS => GSS ... </k> [priority(49)]
+    rule <k> GSS ; .GenStep => GSS ... </k> [priority(49)]
+    rule <k> .GenStep | GSS => GSS ... </k> [priority(49)]
+    rule <k> GSS | .GenStep => GSS ... </k> [priority(49)]
+
+    rule <k> GSS DB:DepthBound => #if DB ==K 0 #then . #else (GSS ; (GSS decrement(DB))) | .GenStep #fi ... </k>
```

Comment:
Perhaps we might also want to simplify `.GenStep DB => .GenStep`

Reply:
Will do.